src/main/java/net/minecraftforge/gui/IGuiProvider.java

```diff
@@ -0,0 +1,62 @@
+package net.minecraftforge.gui;
+
+import io.netty.buffer.ByteBuf;
+import net.minecraft.client.gui.GuiScreen;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.inventory.Container;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.world.World;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents a gui provider- something that activates a gui for a player to use.
+ */
+public interface IGuiProvider<T>
+{
+    /**
+     * Deserializes a buffer from the server into the provider instance.
+     * In practice this is run after an open request to set the owner up on the client.
+     *
+     * @param buffer
+     */
+    IGuiProvider deserialize(ByteBuf buffer);
```

Comment:
In all honesty this return value is highly confusing. And it is using a raw type.